events/nats_subscribe.go

```diff
@@ -0,0 +1,172 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/nats-io/nats.go"
+)
+
+func (c *NatsConnection) coreSubscribe(ctx context.Context, subject string) (<-chan *nats.Msg, error) {
+	logger := c.logger.With(
+		"nats.provider", "core",
+		"nats.subject", subject,
+	)
+
+	sub, err := c.conn.SubscribeSync(subject)
+	if err != nil {
+		return nil, err
+	}
+
+	msgCh := make(chan *nats.Msg, *c.cfg.SubscriberFetchBatchSize)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(msgCh)
+
+				if err := sub.Unsubscribe(); err != nil {
+					logger.Warnw("error unsubscribing", "error", err)
+				}
+
+				return
+			default:
+			}
+
+			if err := c.nextMessage(ctx, sub, msgCh); err != nil {
+				if !errors.Is(err, context.DeadlineExceeded) {
+					logger.Errorw("error fetching messages", "error", err)
+				}
+
+				select {
+				case <-ctx.Done():
+				case <-time.After(*c.cfg.SubscriberFetchDelay):
+				}
+			}
```

Comment:
A `select` like this with a `default` clause doesn't really give us much. Without this, if the context is canceled before reaching `c.nextMessage`, `c.nextMessage` will fail with an appropriate error message (which we will then log) and then we'll select on `<-ctx.Done()` anyway. Putting this logic in the second `select` block will probably simplify things, unless I'm missing something.

Additionally, I'm not clear on whether `SubscriberFetchDelay` should only be the time between fetch attempts in the event of an error or just on all fetches. If it's the latter, the following suggestion should cover everything I'm getting at. I also removed the check for the error type because we probably want to log any errors in fetching messages (and maybe include the subject, which I did not do here).

```suggestion
			if err := c.nextMessage(ctx, sub, msgCh); err != nil {
				logger.Errorw("error fetching messages", "error", err)
			}

			select {
				case <-ctx.Done():
					close(msgCh)

					if err := sub.Unsubscribe(); err != nil {
						logger.Warnw("error unsubscribing", "error", err)
					}

					return
				case <-time.After(*c.cfg.SubscriberFetchDelay):
				}
			}
```

Reply:
yeah, that's a fair point. thanks!

Reply:
Also the fetch delay should only be on error. if we receive data and properly handle it, it makes sense to immediately connect again to get the next chunk of data. but if there's an issue, delaying for a few to ensure there isn't a storm of instance reconnects.

Reply:
Had changed this, but indeed we should pause only when no messages were received. so we end back up with about what we had before, however I did move things which I think make a little easier to read.